velox/adapters/s3fs/S3FileSystem.cpp

```diff
@@ -0,0 +1,267 @@
+  return Aws::String(s.begin(), s.end());
+}
+
+inline std::string_view fromAwsString(const Aws::String& s) {
+  return {s.data(), s.length()};
+}
+
+// Implement the S3ReadFile
+class S3ReadFile final : public ReadFile {
+ public:
+  S3ReadFile(std::string path, std::shared_ptr<Aws::S3::S3Client> client)
+      : path_(std::move(path)), client_(client) {}
+
+  std::string_view pread(uint64_t offset, uint64_t length, Arena* arena)
+      const override {
+    char* pos = arena->reserve(length);
+    preadInternal(offset, length, pos);
+    return {pos, length};
+  }
+
+  std::string_view pread(uint64_t offset, uint64_t length, void* buf)
+      const override {
+    preadInternal(offset, length, static_cast<char*>(buf));
+    return {static_cast<char*>(buf), length};
+  }
+
+  std::string pread(uint64_t offset, uint64_t length) const override {
+    // TODO: use allocator that doesn't initialize memory?
+    std::string result(length, 0);
+    char* pos = result.data();
+    preadInternal(offset, length, pos);
+    return result;
+  }
+
+  uint64_t preadv(
+      uint64_t offset,
+      const std::vector<folly::Range<char*>>& buffers) override {
+    throw std::runtime_error("failure in S3ReadFile::preadv.");
+    return -1;
+  }
+
+  uint64_t size() const override {
+    throw std::runtime_error("failure in S3ReadFile::size.");
+  }
+
+  uint64_t memoryUsage() const override {
+    throw std::runtime_error("failure in S3ReadFile::memoryUsage.");
+    return 0;
+  }
+
+  bool shouldCoalesce() const final {
+    return false;
+  }
+
+ private:
+  void preadInternal(uint64_t offset, uint64_t length, char* pos) const {
+    // Read the desired range of bytes
+    Aws::S3::Model::GetObjectRequest req;
+    Aws::S3::Model::GetObjectResult result;
+    auto first_sep = path_.find_first_of(kSep);
+    req.SetBucket(getAwsString(path_.substr(0, first_sep)));
+    req.SetKey(getAwsString(path_.substr(first_sep + 1)));
+    std::stringstream ss;
+    ss << "bytes=" << offset << "-" << offset + length - 1;
+    req.SetRange(getAwsString(ss.str()));
+    // TODO: Avoid copy below by using  req.SetResponseStreamFactory();
+    // Reference: ARROW-8692
+    auto outcome = client_->GetObject(req);
+    if (!outcome.IsSuccess()) {
+      throw std::runtime_error("failure in S3ReadFile::size preadInternal.");
+    }
+    result = std::move(outcome).GetResultWithOwnership();
+    auto& stream = result.GetBody();
+    stream.read(reinterpret_cast<char*>(pos), length);
+  }
+  std::shared_ptr<Aws::S3::S3Client> client_;
+  std::string path_;
+};
+
+namespace filesystems {
+std::once_flag s3InitializeFlag;
+
+void initializeS3Library() {
+  std::call_once(s3InitializeFlag, []() {
+    Aws::SDKOptions aws_options;
+    aws_options.loggingOptions.logLevel = Aws::Utils::Logging::LogLevel::Fatal;
+    Aws::InitAPI(aws_options);
+  });
+}
+
+namespace S3Config {
+constexpr char const* pathAccessStyle{"hive.s3.path-style-access"};
```

Comment:
why not string_view? It provides a much richer API

Reply:
The use of these constants is purely to act as keys for the Config map and nothing else. So I thought I could avoid the `c_str()` or `data()` calls by using the simple char* instead of string_view.